In [3]:
# Data Processing
import numpy as np
import pandas as pd
from datetime import datetime

#Data Visualizations
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import plotly.express as px 
from scipy import stats

# **Data Understanding**

In [4]:
#load the given data
train_data = pd.read_csv('../data/raw/train.csv',
                        low_memory=False,
                        parse_dates=['Date'],
                        date_parser=(lambda dt: pd.to_datetime(dt, format='%Y-%m-%d')))
test_data = pd.read_csv('../data/raw/test.csv',
                        low_memory=False,
                        parse_dates=['Date'],
                        date_parser=(lambda dt: pd.to_datetime(dt, format='%Y-%m-%d')))
store_data = pd.read_csv('../data/raw/store.csv',
                        low_memory=False,)

C:\Users\USER\AppData\Local\Temp\ipykernel_11056\3003876764.py:2: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  train_data = pd.read_csv('../data/raw/train.csv',
C:\Users\USER\AppData\Local\Temp\ipykernel_11056\3003876764.py:6: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  test_data = pd.read_csv('../data/raw/test.csv',


# **Insights of the training data**

In [5]:
train_data.head() #check the first 5 rows

Store  DayOfWeek       Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5 2015-07-31   5263        555     1      1            0   
1      2          5 2015-07-31   6064        625     1      1            0   
2      3          5 2015-07-31   8314        821     1      1            0   
3      4          5 2015-07-31  13995       1498     1      1            0   
4      5          5 2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [6]:
# checking the informations about the data
train_data.info()
print("\n")
print("The Train dataset has {} Rows and {} Variables".format(str(train_data.shape[0]),str(train_data.shape[1])))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype         
---  ------         --------------    -----         
 0   Store          1017209 non-null  int64         
 1   DayOfWeek      1017209 non-null  int64         
 2   Date           1017209 non-null  datetime64[ns]
 3   Sales          1017209 non-null  int64         
 4   Customers      1017209 non-null  int64         
 5   Open           1017209 non-null  int64         
 6   Promo          1017209 non-null  int64         
 7   StateHoliday   1017209 non-null  object        
 8   SchoolHoliday  1017209 non-null  int64         
dtypes: datetime64[ns](1), int64(7), object(1)
memory usage: 69.8+ MB


The Train dataset has 1017209 Rows and 9 Variables


How many missing feilds each variable has:

In [7]:
train_data.isnull().sum()/train_data.shape[0]

Store            0.0
DayOfWeek        0.0
Date             0.0
Sales            0.0
Customers        0.0
Open             0.0
Promo            0.0
StateHoliday     0.0
SchoolHoliday    0.0
dtype: float64

- We see that the training data doesn't have any nulls

In [8]:
print("The Date range in the data is from {} to {}".format(train_data.Date.min(),train_data.Date.max()))

The Date range in the data is from 2013-01-01 00:00:00 to 2015-07-31 00:00:00


In [9]:
train_data.shape

(1017209, 9)

- So, the data in the training set is from Jan 1st, 20153to July 31st, 2015 

In [10]:
# Checking if the sales are zero in the data
train_data[train_data.Sales == 0.0]

Store  DayOfWeek       Date  Sales  Customers  Open  Promo  \
291        292          5 2015-07-31      0          0     0      1   
875        876          5 2015-07-31      0          0     0      1   
1406       292          4 2015-07-30      0          0     0      1   
1990       876          4 2015-07-30      0          0     0      1   
2521       292          3 2015-07-29      0          0     0      1   
...        ...        ...        ...    ...        ...   ...    ...   
1017204   1111          2 2013-01-01      0          0     0      0   
1017205   1112          2 2013-01-01      0          0     0      0   
1017206   1113          2 2013-01-01      0          0     0      0   
1017207   1114          2 2013-01-01      0          0     0      0   
1017208   1115          2 2013-01-01      0          0     0      0   

        StateHoliday  SchoolHoliday  
291                0              1  
875                0              1  
1406               0              1  
1990               0              1  
2521               0              1  
...              ...            ...  
1017204            a              1  
1017205            a              1  
1017206            a              1  
1017207            a              1  
1017208            a              1  

[172871 rows x 9 columns]

- From total 1017209 rows, almost on 172871 rows the sales was zero. So, lets discover why is that.

# Exploratory Data Analysis(EDA)

In [13]:
# since it would be cluttered to plot sales against date for all store so I will pick a random store for the sake of visualisations

store_1111_data = train_data[train_data["Store"] == 1111]
fig = px.scatter(store_1111_data, x="Date" , y="Sales",
             hover_data={"Date": "|%B %d,%Y"},color="Open",
             title='Daily Sales for Store 1111(store was open/closed)')
fig.update_xaxes(dtick="M1",tickformat="%b\n%Y")
fig.update_traces(marker=dict(color=['green' if i==1 else 'red' for i in store_1111_data['Open']]))
fig.show()